File: apps/lantern-api/src/features/home/presentation/http/handle_theme.py
from __future__ import annotations

import json
import time
import uuid

from fireweave import EvaluationContext

from src.core.runtime_context import RuntimeContext
from src.features.home.application.pick_theme import pick_theme
from src.features.telemetry.infrastructure.start_otel import Telemetry
from src.fireweave.fw_harness import fw_control_points

_INSTANCE_KEY = f"inst_{uuid.uuid4().hex}"


def handle_theme(
    ctx: RuntimeContext,
    telemetry: Telemetry,
    targeting_key: str | None,
) -> tuple[int, bytes]:
    started = time.perf_counter()
    key = targeting_key or _INSTANCE_KEY
    # @fireweave-controlpoint random-theme
    enabled = fw_control_points().get_boolean_value(
        "random-theme",
        False,
        EvaluationContext(targeting_key=key),
    )
    if not enabled:
        return 404, json.dumps({"error": "theme off", "service": ctx.service}).encode()
    theme = pick_theme()
    telemetry.increment("random-theme.served")
    telemetry.record("random-theme.latency_ms", (time.perf_counter() - started) * 1000)
    return 200, json.dumps({"service": ctx.service, "theme": theme}).encode()

File: apps/atlas-api/src/features/home/presentation/http/handle-theme.ts
import type { RuntimeContext } from "../../../../core/runtime-context";
import type { Telemetry } from "../../../telemetry/application/ports/telemetry";
import { fw } from "../../../../fireweave/fw-harness";
import { pickTheme } from "../../application/pick-theme";

export async function handleTheme(
  ctx: RuntimeContext,
  telemetry: Telemetry,
  targetingKey: string,
): Promise<Response> {
  const started = performance.now();
  // @fireweave-controlpoint random-theme
  const enabled = await fw.controlPoints.getBooleanValue("random-theme", false, {
    targetingKey,
  });
  if (!enabled) {
    return Response.json({ error: "theme off", service: ctx.service }, { status: 404 });
  }
  const theme = pickTheme();
  telemetry.increment("random-theme.served", { theme: theme.name });
  telemetry.record("random-theme.latency_ms", performance.now() - started);
  return Response.json({ service: ctx.service, theme });
}
